File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/scenario/actions/SEAction.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.scenario.actions;

import java.io.Serializable;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import mil.tatrc.physiology.datamodel.bind.ActionData;
import mil.tatrc.physiology.datamodel.properties.SEScalarTime;
import mil.tatrc.physiology.datamodel.substance.SESubstance;
import mil.tatrc.physiology.datamodel.substance.SESubstanceCompound;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;
import mil.tatrc.physiology.utilities.Log;

public abstract class SEAction implements Serializable
{
	protected String comment;
	protected SEScalarTime scenarioTime;
	
	
	public SEAction() 
	{
		comment = null;
		scenarioTime = null;
	}
	
	public void copy(SEAction other) 
	{
		reset();
		comment = other.comment;
	}
	
	public void reset() 
	{
		comment = null;
	}
	
	public boolean load(ActionData in) 
	{
		reset();		
		this.comment = in.getComment();
		return true;
	}
	
	public abstract ActionData unload();
	
	protected void unload(ActionData data)
	{
		if(hasComment())
			data.setComment(this.comment);
	}
	
	public boolean hasComment()
	{
		return this.comment != null && !this.comment.isEmpty();
	}	
	public String getComment()
	{
		return this.comment;
	}
	public void setComment(String comment)
	{
		this.comment = comment;
	}
	public void invalidateComment()
	{
		this.comment = null;
	}
	
	
	public boolean hasScenarioTime()
	{
		return this.scenarioTime == null ? false : this.scenarioTime.isValid();
	}	
	public SEScalarTime getScenarioTime()
	{
		if(this.scenarioTime == null)
			this.scenarioTime = new SEScalarTime();
		return this.scenarioTime;
	}
	
	
	protected static List<String> actionPackages;
	public synchronized static SEAction bind2CDM(Object bind, SESubstanceManager subs)
	{
		String cClassName="";

		Class<?>  c=null;
		String bClassName = bind.getClass().getName();
		
		if(actionPackages==null)
		{
			actionPackages = new ArrayList<String>();
			actionPackages.add("scenario.actions");
			actionPackages.add("patient.actions");
			actionPackages.add("system.equipment.anesthesia.actions");
			actionPackages.add("system.equipment.inhaler.actions");
			actionPackages.add("system.environment.actions");
		}

		for(String pkg : actionPackages)
		{
			c=null;
			cClassName=bClassName.replace("bind", pkg);
			int s = cClassName.lastIndexOf(".");
			int e = cClassName.indexOf("Data", s);
			cClassName = cClassName.substring(0, s+1) + "SE" + cClassName.substring(s+1,e);
			try{
				c = Class.forName(cClassName);
				break;
			}catch(Exception ex){};
		}
		try
		{
			if(c!=null)
			{
				Object o=null;
				for(Constructor<?> ctor : c.getConstructors())
				{
					Class<?>[] params = ctor.getParameterTypes();
					if(params.length>0)
					{
						if(params[0].isAssignableFrom(SESubstance.class))
						{
							Method m=bind.getClass().getMethod("getSubstance");
							String subName = (String)m.invoke(bind);
							o=ctor.newInstance(subs.getSubstance(subName));
						}
						else if(params[0].isAssignableFrom(SESubstanceCompound.class))
						{
							Method m=bind.getClass().getMethod("getSubstanceCompound");
							String subName = (String)m.invoke(bind);
							o=ctor.newInstance(subs.getCompound(subName));
						}
					}
				}
				if(o==null)
					o = c.newInstance();
				for(Method m : c.getMethods())
				{
					if(m.getName().equals("load"))
					{
						if(m.getParameterTypes().length==1)						
							m.invoke(o, bind);
						else// I assume you need the substance manager, which is a good assumption now...
							m.invoke(o, bind, subs);
						return (SEAction)o;
					}
				}
			}
			Log.error("SE Class not found for "+bClassName);
		}
		catch(Exception ex)
		{
			Log.error("Error looking for CDM class for "+bClassName,ex);
		}
		return null;
	}
	
	public abstract boolean isValid();
	public abstract String toString();
}

File: projects/biogears/libBiogears/src/cdm/scenario/SEAction.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/scenario/SEAction.h>

#include "io/cdm/Actions.h"
#include "io/cdm/AnesthesiaActions.h"
#include "io/cdm/EnvironmentActions.h"
#include "io/cdm/EnvironmentConditions.h"
#include "io/cdm/InhalerActions.h"
#include "io/cdm/PatientActions.h"
#include "io/cdm/PatientAssessments.h"

#include <biogears/cdm/patient/actions/SEAcuteRespiratoryDistress.h>
#include <biogears/cdm/patient/actions/SEAcuteStress.h>
#include <biogears/cdm/patient/actions/SEAirwayObstruction.h>
#include <biogears/cdm/patient/actions/SEApnea.h>
#include <biogears/cdm/patient/actions/SEAsthmaAttack.h>
#include <biogears/cdm/patient/actions/SEBrainInjury.h>
#include <biogears/cdm/patient/actions/SEBronchoconstriction.h>
#include <biogears/cdm/patient/actions/SEBurnWound.h>
#include <biogears/cdm/patient/actions/SECardiacArrest.h>
#include <biogears/cdm/patient/actions/SEChestCompression.h>
#include <biogears/cdm/patient/actions/SEChestCompressionForce.h>
#include <biogears/cdm/patient/actions/SEChestCompressionForceScale.h>
#include <biogears/cdm/patient/actions/SEChestOcclusiveDressing.h>
#include <biogears/cdm/patient/actions/SEConsciousRespiration.h>
#include <biogears/cdm/patient/actions/SEConsumeNutrients.h>
#include <biogears/cdm/patient/actions/SEEbola.h>
#include <biogears/cdm/patient/actions/SEEscharotomy.h>
#include <biogears/cdm/patient/actions/SEExampleAction.h>
#include <biogears/cdm/patient/actions/SEExercise.h>
#include <biogears/cdm/patient/actions/SEHemorrhage.h>
#include <biogears/cdm/patient/actions/SEInfection.h>
#include <biogears/cdm/patient/actions/SEIntubation.h>
#include <biogears/cdm/patient/actions/SEMechanicalVentilation.h>
#include <biogears/cdm/patient/actions/SENasalCannula.h>
#include <biogears/cdm/patient/actions/SENeedleDecompression.h>
#include <biogears/cdm/patient/actions/SEOverride.h>
#include <biogears/cdm/patient/actions/SEPainStimulus.h>
#include <biogears/cdm/patient/actions/SEPatientAssessmentRequest.h>
#include <biogears/cdm/patient/actions/SEPericardialEffusion.h>
#include <biogears/cdm/patient/actions/SEPulmonaryShunt.h>
#include <biogears/cdm/patient/actions/SERadiationAbsorbedDose.h>
#include <biogears/cdm/patient/actions/SESleep.h>
#include <biogears/cdm/patient/actions/SESubstanceBolus.h>
#include <biogears/cdm/patient/actions/SESubstanceCompoundInfusion.h>
#include <biogears/cdm/patient/actions/SESubstanceInfusion.h>
#include <biogears/cdm/patient/actions/SESubstanceNasalDose.h>
#include <biogears/cdm/patient/actions/SESubstanceOralDose.h>
#include <biogears/cdm/patient/actions/SETensionPneumothorax.h>
#include <biogears/cdm/patient/actions/SETourniquet.h>
#include <biogears/cdm/patient/actions/SEUrinate.h>
#include <biogears/cdm/scenario/SEAdvanceTime.h>
#include <biogears/cdm/scenario/SESerializeState.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/cdm/system/environment/actions/SEEnvironmentChange.h>
#include <biogears/cdm/system/environment/actions/SEThermalApplication.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEAnesthesiaMachineConfiguration.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEExpiratoryValveLeak.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEExpiratoryValveObstruction.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEInspiratoryValveLeak.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEInspiratoryValveObstruction.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEMaskLeak.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEOxygenTankPressureLoss.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEOxygenWallPortPressureLoss.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SESodaLimeFailure.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SETubeCuffLeak.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEVaporizerFailure.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEVentilatorPressureLoss.h>
#include <biogears/cdm/system/equipment/Anesthesia/actions/SEYPieceDisconnect.h>
#include <biogears/cdm/system/equipment/Inhaler/actions/SEInhalerConfiguration.h>

namespace std {
template class vector<biogears::SEAction*>;
}

namespace biogears {
SEAction::SEAction(Logger* logger)
  : Loggable(logger)
{
  m_Comment = "";
}
//-----------------------------------------------------------------------------
SEAction::~SEAction()
{
  Clear();
}
//-----------------------------------------------------------------------------
void SEAction::Clear()
{
  m_Comment = "";
}
//-----------------------------------------------------------------------------
const char* SEAction::GetComment() const
{
  return m_Comment.c_str();
}
//-----------------------------------------------------------------------------
void SEAction::SetComment(const char* comment)
{
  m_Comment = comment;
}
//-----------------------------------------------------------------------------
void SEAction::SetComment(const std::string& comment)
{
  m_Comment = comment;
}
//-----------------------------------------------------------------------------
bool SEAction::HasComment() const
{
  return m_Comment.empty() ? false : true;
}
//-----------------------------------------------------------------------------
void SEAction::InvalidateComment()
{
  m_Comment = "";
}
//-----------------------------------------------------------------------------
const char* SEAction::ToString() const
{
  std::stringstream ss;
  this->ToString(ss);
  m_StringRepresnetation = ss.str();
  return m_StringRepresnetation.c_str();
}
//-----------------------------------------------------------------------------
//-------------------------------------------------------------------------------
bool SEAction::operator==(const SEAction& rhs) const
{
  return m_Comment == rhs.m_Comment;
}
bool SEAction::operator!=(const SEAction& rhs) const
{
  return !(*this == rhs);
}
//-------------------------------------------------------------------------------
}